driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -326,14 +332,43 @@ private void logServerSelectionFailure(final ServerSelector serverSelector, fina
         }
     }
 
+    @Nullable
+    private ServerTuple selectServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription) {
+        return selectServer(serverSelector, clusterDescription, this::getServer, getRandom());
+    }
 
-    private ServerTuple selectRandomServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription) {
-        List<ServerDescription> serverDescriptions = serverSelector.select(clusterDescription);
-        if (!serverDescriptions.isEmpty()) {
-            return getRandomServer(new ArrayList<ServerDescription>(serverDescriptions));
-        } else {
-            return null;
+    @Nullable
+    @VisibleForTesting(otherwise = PRIVATE)
+    static ServerTuple selectServer(final ServerSelector serverSelector, final ClusterDescription clusterDescription,
+            final Function<ServerAddress, Server> serverCatalog, final Random random) {
+        return atMostNRandom(new ArrayList<>(serverSelector.select(clusterDescription)), 2, random, serverDescription -> {
```

Comment:
Nit: It's unfortunate that we have to make a copy of the List, but I don't see a way around it.  But since the requirement that the list is RandomAccess is internal to the atMostNRandom method, maybe it would be clearer if atMostNRandom was responsible for copying the list?

Reply:
If `atMostNRandom` makes the copying itself, then it makes calling `atMostNRandom` more than once per list ineffective (not that we are doing so), and also, copying does not seem like the job relevant to the method. I left the decision to copy or not up to the caller.